## Data Loading

In [ ]:
!pip install python-dotenv google-api-python-client

from google.colab import drive
drive.mount('/content/drive')

import os
from dotenv import load_dotenv
load_dotenv("/content/drive/MyDrive/Professional/Portfolio/test_split/.env_X_train")
load_dotenv("/content/drive/MyDrive/Professional/Portfolio/test_split/.env_X_test")
load_dotenv("/content/drive/MyDrive/Professional/Portfolio/test_split/.env_y_train")
load_dotenv("/content/drive/MyDrive/Professional/Portfolio/test_split/.env_y_test")
x_train_id = os.getenv("X_TRAIN_ID")
x_test_id = os.getenv("X_TEST_ID")
y_train_id = os.getenv("Y_TRAIN_ID")
y_test_id = os.getenv("Y_TEST_ID")

load_dotenv("/content/drive/MyDrive/Professional/Portfolio/test_split/.env_github")
github_pat = os.getenv("GITHUB_PAT")

if not x_train_id or not x_test_id or not y_train_id or not y_test_id:
    raise ValueError("❌ Error: One or more environment variables are missing or invalid.")
if not github_pat:
    raise ValueError("❌ Error: 'GITHUB_PAT' is missing or invalid in your .env file.")

!git clone https://{github_pat}@github.com/vmagdale2/data-split-model-performance-analysis.git

import sys
sys.path.append('/content/data-split-model-performance-analysis/scripts')
%cd /content/data-split-model-performance-analysis/scripts
!pwd
!ls

from utils import authenticate_and_load_env, load_data_from_drive

service = authenticate_and_load_env()

x_train_id = os.getenv("X_TRAIN_ID")
x_test_id = os.getenv("X_TEST_ID")
y_train_id = os.getenv("Y_TRAIN_ID")
y_test_id = os.getenv("Y_TEST_ID")

df_X_train = load_data_from_drive(service, x_train_id)
df_X_test = load_data_from_drive(service, x_test_id)
df_y_train = load_data_from_drive(service, y_train_id)
df_y_test = load_data_from_drive(service, y_test_id)

print("✅ Data loaded successfully!")

display(df_X_train.head())
display(df_X_test.head())
display(df_y_train.head())
display(df_y_test.head())

Mounted at /content/drive
Cloning into 'data-split-model-performance-analysis'...
remote: Enumerating objects: 87, done.
remote: Counting objects: 100% (87/87), done.
remote: Compressing objects: 100% (61/61), done.
remote: Total 87 (delta 38), reused 66 (delta 20), pack-reused 0 (from 0)
Receiving objects: 100% (87/87), 296.58 KiB | 4.71 MiB/s, done.
Resolving deltas: 100% (38/38), done.
/content/data-split-model-performance-analysis/scripts
/content/data-split-model-performance-analysis/scripts
data_preprocessing.py  readme.md	    train_test_split.py  visualization.py
modeling.py	       scaling_analysis.py  utils.py
✅ Google Drive API authenticated successfully!
✅ Data loaded successfully!
✅ Data loaded successfully!
✅ Data loaded successfully!
✅ Data loaded successfully!
✅ Data loaded successfully!


1stFlrSF  2ndFlrSF  3SsnPorch  BedroomAbvGr  BsmtFinSF1  BsmtFinSF2  \
0     630.0       0.0        0.0             1       515.0         0.0   
1     845.0       0.0        0.0             3         0.0         0.0   
2     728.0     728.0        0.0             3         0.0         0.0   
3     561.0     668.0        0.0             2       285.0         0.0   
4    1601.0       0.0        0.0             3      1358.0         0.0   

   BsmtFullBath  BsmtHalfBath  BsmtUnfSF  EnclosedPorch  ...  SaleType_CWD  \
0             1             0      115.0            0.0  ...         False   
1             0             0        0.0            0.0  ...         False   
2             0             0      728.0            0.0  ...         False   
3             0             0      276.0            0.0  ...         False   
4             1             0      223.0            0.0  ...         False   

   SaleType_Con  SaleType_ConLD  SaleType_ConLI  SaleType_ConLw  SaleType_New  \
0         False           False           False           False         False   
1         False           False           False           False         False   
2         False           False           False           False         False   
3         False           False           False           False         False   
4         False           False           False           False         False   

   SaleType_Oth  SaleType_WD  Street_Pave  Utilities_NoSeWa  
0         False         True         True             False  
1         False         True         True             False  
2         False         True         True             False  
3         False         True         True             False  
4         False         True         True             False  

[5 rows x 240 columns]

1stFlrSF  2ndFlrSF  3SsnPorch  BedroomAbvGr  BsmtFinSF1  BsmtFinSF2  \
0    1518.0       0.0        0.0             1      1218.0         0.0   
1     925.0       0.0        0.0             2       338.0       466.0   
2    1095.0     679.0        0.0             4         0.0         0.0   
3     888.0     868.0        0.0             3       742.0         0.0   
4    1337.0       0.0        0.0             3       699.0         0.0   

   BsmtFullBath  BsmtHalfBath  BsmtUnfSF  EnclosedPorch  ...  SaleType_CWD  \
0             0             0      300.0            0.0  ...         False   
1             0             1      121.0            0.0  ...         False   
2             1             0     1095.0           90.0  ...         False   
3             1             0      130.0            0.0  ...         False   
4             1             0      638.0            0.0  ...         False   

   SaleType_Con  SaleType_ConLD  SaleType_ConLI  SaleType_ConLw  SaleType_New  \
0         False           False           False           False         False   
1         False           False           False           False         False   
2         False           False           False           False         False   
3         False           False           False           False         False   
4         False           False           False           False         False   

   SaleType_Oth  SaleType_WD  Street_Pave  Utilities_NoSeWa  
0         False         True         True             False  
1         False         True         True             False  
2         False         True         True             False  
3         False         True         True             False  
4         False         True         True             False  

[5 rows x 240 columns]

SalePrice
0    86000.0
1    84000.0
2   176000.0
3   124000.0
4   272000.0

SalePrice
0   274000.0
1   117500.0
2    87000.0
3   204000.0
4   185000.0

In [35]:
import pandas as pd
from sklearn.preprocessing import RobustScaler
from sklearn.linear_model import Ridge
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import joblib
import matplotlib.pyplot as plt
import numpy as np
import numpy as np
import pandas as pd
from sklearn.linear_model import Ridge
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import joblib

In [34]:
X_train = pd.DataFrame(df_X_train)
X_test = pd.DataFrame(df_X_test)
y_train = pd.DataFrame(df_y_train).values.ravel()
y_test = pd.DataFrame(df_y_test).values.ravel()

----------

## Objectives
- Perform advanced hyperparameter tuning to improve model performance.
- Use GridSearchCV to optimize Ridge Regression parameters.
- Compare tuned Ridge model performance with previous results.

# Hyperparameter Tuning

## Tuning

In [43]:
from sklearn.preprocessing import RobustScaler

scaler = RobustScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [44]:
param_grid = {'alpha': np.logspace(-2, 3, 10)}
ridge_grid = GridSearchCV(Ridge(), param_grid, cv=5)
ridge_grid.fit(X_train_scaled, y_train)

GridSearchCV(cv=5, estimator=Ridge(),
             param_grid={'alpha': array([1.00000000e-02, 3.59381366e-02, 1.29154967e-01, 4.64158883e-01,
       1.66810054e+00, 5.99484250e+00, 2.15443469e+01, 7.74263683e+01,
       2.78255940e+02, 1.00000000e+03])})

In [49]:
best_ridge_model = ridge_grid.best_estimator_
print(f"✅ Best Alpha: {ridge_grid.best_params_['alpha']}")
y_pred = best_ridge_model.predict(X_test_scaled)

rmse = np.sqrt(mean_squared_error(y_test, y_pred))
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Root Mean Squared Error (RMSE): {rmse:.2f}")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"R-squared (R²): {r2:.2f}")

✅ Best Alpha: 21.544346900318846
Root Mean Squared Error (RMSE): 32644.34
Mean Absolute Error (MAE): 19036.73
R-squared (R²): 0.82


In [50]:
y_pred = best_ridge_model.predict(X_test_scaled)
y_pred_df = pd.DataFrame(y_pred, columns=['Predicted_Values'])
y_pred_df.to_csv('/content/drive/My Drive/Professional/Portfolio/test_split/y_pred.csv', index=False)
joblib.dump(best_ridge_model, '/content/drive/My Drive/Professional/Portfolio/test_split/best_ridge_model.pkl')

['/content/drive/My Drive/Professional/Portfolio/test_split/best_ridge_model.pkl']

# Fine-Tuned Ridge Regression Results

## 📊 Overview
This document presents the results of the fine-tuned Ridge Regression model from the **`04_fine_tuning.ipynb`** notebook. The improved model performance reflects effective tuning and proper scaling techniques.

---

## ⚙️ Model Details
- **Algorithm:** Ridge Regression
- **Scaling Method:** RobustScaler (Improved stability in presence of outliers)
- **Best Alpha (Tuned Parameter):** `21.5443`
- **Cross-Validation:** 5-Fold GridSearchCV

---

## 📈 Performance Metrics
| **Metric**                | **Value**        |
|---------------------------|------------------|
| **Root Mean Squared Error (RMSE)** | 32,644.34     |
| **Mean Absolute Error (MAE)**      | 19,036.73     |
| **R-squared (R²)**                | 0.82            |

### 🔎 **Interpretation:**
✅ **R² = 0.82** — A strong fit, indicating that 82% of the variance in the target variable is explained by the model.  
✅ **RMSE = 32,644** — Represents an average prediction error of about 32K. This is reasonable given the variability in housing prices.  
✅ **MAE = 19,036** — Reflects an average absolute error of ~$19K per prediction, showing improved precision.




